### PR TITLE
Dual shape function modification

### PR DESCRIPTION
**Pull request by dewenyushu**
This is a re-opening of Jacob's draft PR on an initial implementation of dual basis modification for 3D contact and edge dropping (#3023)

Improvement on the efficiency and robustness of dual basis calculation will be examined in separate PRs.

### CODE REVIEW

**Comment by dewenyushu** [comment]
@lindsayad @jbadger95 I made some changes to the dual basis calculation PR. It's now failing one test in stochastic tools somehow. I will continue to look into it. Aside from it, this should be ready for your review. 

I added back the calculation of dual shape coefficient using the default QRule, which can be turned off on the MOOSE level for msm elements that are partially in contact. This way we can protect the calculation of the dual shape function without dual coefficients being calculated first. Also since we _may_ still need the default dual shape function for the interior msm elements I suppose it's a better idea to retain this capability in libmesh.

**Comment by jwpeterson** [comment]
> It's now failing one test in stochastic tools

That same test is failing on lots of PRs, so probably not related to this one.

**Comment by jbadger95** [comment]
Have you looked into the cost or conditioning at all?

**Comment by dewenyushu** [comment]
> Have you looked into the cost or conditioning at all?

I looked into the conditioning. There's significant increase of condition number (several orders of magnitude) when the element in contact becomes small (see the attached plot, condition number shoots up when the secondary element is about to be fully not in contact). I am working on an initial implementation of element/nodal scaling for improving this behavior. 

[Condition_Number_2D_Slide_Scaling.pdf](https://github.com/libMesh/libmesh/files/7464090/Condition_Number_2D_Slide_Scaling.pdf)

I have not looked deep into the cost yet. I suppose we have not come into a conclusion whether we will need customized quadrature points for the interior elements, so it is not very clear to me right now in terms of where and how to reduce the cost. 

Anyways, my intent is to take smaller steps to cleanup and merge what you've implemented first and optimize things (better scaling & improve efficiency) later, since I realize both tasks would take some more work than I expected.

**Comment by jbadger95** [comment]
Is this the conditioning of the overall system or of the mass matrix for the dual coefficients?

(sorry for the delayed response, busy day)

**Comment by jbadger95** [comment]
My biggest concern with conditioning is that on 3D hexas (quad faces) that if we don’t enrich the quadrature order then the dual mass matrix can potentially be singular. The poor conditioning of edge dropping is expected (without scaling), my concern here is just that for normal mortar problems (without edge dropping) using an inexact quadrature might cause problems.

Can you compare the maximum conditioning number for dual mass matrices (on an aligned 3D hexa mesh and maybe some twisted meshes or something with a bit more exotic mortar segment mesh) using the default mortar quadrature rule vs an enriched quadrature rule (or the default primal quadrature rule if you wish)?  

Do all the MOOSE tests pass with these changes?

**Comment by dewenyushu** [comment]
> Is this the conditioning of the overall system or of the mass matrix for the dual coefficients?
> 
> (sorry for the delayed response, busy day)

This is the conditioning of the overall system.

**Comment by dewenyushu** [comment]
> Can you compare the maximum conditioning number for dual mass matrices (on an aligned 3D hexa mesh and maybe some twisted meshes or something with a bit more exotic mortar segment mesh) using the default mortar quadrature rule vs an enriched quadrature rule (or the default primal quadrature rule if you wish)?

Thanks Jacob!

Just to make sure, by `enriched qrule` you mean the customized qrule computed in `ComputeMortarFunctor` in this PR (https://github.com/idaholab/moose/pull/18885), or do you mean a higher quadrature order is needed while computing the customized qrule?

> Do all the MOOSE tests pass with these changes?

Yes, all the MOOSE tests pass (except for the one stochastic tools that does not seem to be related).

**Comment by jbadger95** [comment]
By enriched quadrature I mean an enriched mortar quadrature (e.g. `quadrature = SECOND` In mortar constraint). A first order quadrature on the triangle mortar segments does not exactly integrate first order quad elements, that’s why I’m worried about dual mass matrices being poorly conditioned.

**Comment by dewenyushu** [comment]
> By enriched quadrature I mean an enriched mortar quadrature (e.g. `quadrature = SECOND` In mortar constraint). A first order quadrature on the triangle mortar segments does not exactly integrate first order quad elements, that’s why I’m worried about dual mass matrices being poorly conditioned.

Okay, I got what you mean here.

I tested this change (together with your MOOSE change) with two non-conforming blocks (hex 8 elements), default qrule, with and without edge dropping. Both work fine. 

I can try to get the condition number of the mass matrix and compare that for default and enriched quadrature to determine if enriched qrule is needed.

**Comment by jbadger95** [comment]
Sounds good, sorry for pestering you I just didn’t get a chance to really test it so just wanted to make sure you had. If you’re convinced it’s not breaking anything I don’t have any objections to merging it. I’ll review it a bit later today.

**Comment by dewenyushu** [comment]
> Sounds good, sorry for pestering you I just didn’t get a chance to really test it so just wanted to make sure you had. If you’re convinced it’s not breaking anything I don’t have any objections to merging it. I’ll review it a bit later today.

Not at all! Glad that we are on the same page. And I really appreciate your help looking into this.

File: src/mesh/unstructured_mesh.C
Comment (roystgnr) [inline src/mesh/unstructured_mesh.C:103]: This seems like an important and distinct change, doesn't it?  Probably worth backporting?  Is it worth merging separately if this PR gets delayed any further?

File: include/fe/fe_base.h
Comment (roystgnr) [inline include/fe/fe_base.h:592]: The documentation here is an atavism now, yes?

File: src/fe/fe_base.C
Comment (roystgnr) [inline src/fe/fe_base.C:747]: This seems to be the key to the whole commit here, yet this sentence is too cryptic for me to understand it.

File: include/fe/fe_base.h
Comment (roystgnr) [inline include/fe/fe_base.h:798]: As long as there are more changes to make anyway, let's just turn this commit into a fixup on top of the compilation-breaking commit it fixed?

File: include/fe/fe_abstract.h
Comment (roystgnr) [inline include/fe/fe_abstract.h:162]: Likewise with a fixup for this commit, if I'm understanding correctly and it broke --enable-ifem builds.

File: src/mesh/unstructured_mesh.C
Comment (dewenyushu) [reply src/mesh/unstructured_mesh.C:103]: Yeah, this is not closely related to this PR. Let me have it in a separate one.

File: include/fe/fe_base.h
Comment (dewenyushu) [reply include/fe/fe_base.h:592]: Good catch. The docstring is outdated. Let me update it

File: src/fe/fe_base.C
Comment (dewenyushu) [reply src/fe/fe_base.C:747]: Sure, let me clarify things a bit more.

File: src/mesh/unstructured_mesh.C
Comment (jbadger95) [reply src/mesh/unstructured_mesh.C:103]: This was actually a @lindsayad change, [#3013](https://github.com/libMesh/libmesh/pull/3013) (sorry for including it here, I figured it would resolve on rebase anyway)

**Comment by dewenyushu** [comment]
Okay, I check on the conditioning of the mass matrix using FIRST and SECOND quadrature. I pick the maximum condition number in all msm element throughout the simulation. This is done for three cases:

****Case 1**, two block compression**
Maximum condition number:
FIRST order -  8.73517; SECOND order - 8.73517


****Case 2**, two block compression, rotated top block** 
Maximum condition number:
FIRST order -  16.7234; SECOND order - 16.308



****Case 3**, two block sliding**
Maximum condition number:
FIRST order -  236.946; SECOND order - 235.013 (maximum condition number at around the point when one element is to be fully not in contact, and then reduces to ~3 in both cases)

**Comment by jbadger95** [comment]
Awesome, that’s interesting how little it affects the conditioning!

**Review by jbadger95 — APPROVED** [review]
Looks good to me, thanks for finishing this up!

File: include/fe/fe_abstract.h
Comment (lindsayad) [inline include/fe/fe_abstract.h:171]: Do not all classes go through FE? Just wondering why this is a run time error as opposed to pure virtual

File: src/fe/fe_base.C
Comment (lindsayad) [inline src/fe/fe_base.C:753]: const

File: include/fe/fe_abstract.h
Comment (lindsayad) [reply include/fe/fe_abstract.h:171]: Ah I see that InfFE does not. Still I wonder if it would be nicer to have this be pure virtual and put the not implemented error in InfFE? I don’t know what the style preference is for @jwpeterson and @roystgnr

Comment (jwpeterson) [reply include/fe/fe_abstract.h:171]: I don't have a strong feeling either way for FE since as you mentioned the only other implementation is InfFE

Comment (roystgnr) [reply include/fe/fe_abstract.h:171]: I'm happy either way.  We tend to make our most "critical" methods pure virtual, to try to set an example to subclass authors that implementation is something only to be avoided if you can't help it, but the definition of "critical" is a matter of opinion, and especially in this case where we're not expecting any completely new subclasses any time soon the matter seems just academic.

Comment (dewenyushu) [reply include/fe/fe_abstract.h]: Thanks @lindsayad @jwpeterson  @roystgnr for your inputs - Let me make this as pure virtual and add the not implemented error in InfFE

**Comment by dewenyushu** [comment]
I cannot see the details in Civet test for the Coverage for some reason. Could someone help me look into it? - Is it because the functions in InfFE are not tested?

File: include/fe/fe_abstract.h
Comment (dewenyushu) [reply include/fe/fe_abstract.h]: Moving it to InfFE does seem to cause some extra test issue. I could not see the details in Civet to make sure whether this is only coverage though. I'll just keep this inside of `fe_base` instead

File: src/fe/fe.C
Comment (jwpeterson) [inline src/fe/fe.C:305]: This comment is referring to MOOSE code "Assembly::reinitDual" and MOOSE jargon "msm". If you would like to keep it here, maybe just clarify that is the case?

**Review by jwpeterson — COMMENTED** [review]
I don't know enough about the details of this PR to be a final Approver, @lindsayad please merge this when you are ready.

File: include/fe/fe_abstract.h
Comment (lindsayad) [reply include/fe/fe_abstract.h]: The coverage error has been present for every single completed testing of this PR: https://civet.inl.gov/pr/14815/ so the issue is not this single function. I suspect the orange box is due to the fact that the coverage % of this PR is less than the suggested 90%

Comment (lindsayad) [reply include/fe/fe_abstract.h]: next time let's not force push so quickly

Comment (lindsayad) [reply include/fe/fe_abstract.h]: Yea this is from the coverage testing script:
```
          if coverage_passes $COVERAGE_GENERATE_DIFF_NEW_RATE $WARN_NEW; then
            PASSED_COVERAGE+=("New line coverage rate $COVERAGE_GENERATE_DIFF_NEW_RATE%% passes the suggested $WARN_NEW%%")
          else
            WARNED_COVERAGE+=("New new line coverage rate $COVERAGE_GENERATE_DIFF_NEW_RATE%% is less than the suggested $WARN_NEW%%")
            RETURN_CODE=1
```

Comment (lindsayad) [reply include/fe/fe_abstract.h]: @jwpeterson was right ... even though the timing was very suspiciously correlated with the move to InfFE and back it was not caused by it. @loganharbour pushed a commit fixing the storage logic yesterday afternoon (it had broken last Thursday)

Comment (dewenyushu) [reply include/fe/fe_abstract.h]: Thanks for checking on this! It was such a coincidence that my update was at the same time frame as Logan's.

**Pull request merged**
